File: src/main/java/myproject/study/books_store/service/AIService.java
package myproject.study.books_store.service;

import myproject.study.books_store.dto.AIChatRequest;
import myproject.study.books_store.dto.AIChatResponse;
import myproject.study.books_store.dto.AISearchRequest;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

/**
 * Service for interacting with Python AI Service
 */
@Service
public class AIService {

    // @Value("${ai.service.url:http://localhost:8000}")
    @Value("${ai.service.url}")
    private String aiServiceUrl;

    private final RestTemplate restTemplate;

    public AIService() {
        this.restTemplate = new RestTemplate();
    }

    /**
     * Send chat message to AI service
     */
    public AIChatResponse chat(AIChatRequest request) {
        String url = aiServiceUrl + "/api/chat";

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        HttpEntity<AIChatRequest> entity = new HttpEntity<>(request, headers);

        return restTemplate.postForObject(url, entity, AIChatResponse.class);
    }

    /**
     * Semantic search for books
     */
    public Object search(AISearchRequest request) {
        String url = aiServiceUrl + "/api/search";

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        HttpEntity<AISearchRequest> entity = new HttpEntity<>(request, headers);

        return restTemplate.postForObject(url, entity, Object.class);
    }

    /**
     * Get similar books
     */
    public Object getSimilarBooks(Long bookId, Integer topK) {
        String url = aiServiceUrl + "/api/similar";

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        var request = new SimilarBooksRequest(bookId, topK);
        HttpEntity<SimilarBooksRequest> entity = new HttpEntity<>(request, headers);

        return restTemplate.postForObject(url, entity, Object.class);
    }

    /**
     * Check AI service health
     */
    public Object checkHealth() {
        String url = aiServiceUrl + "/health";
        return restTemplate.getForObject(url, Object.class);
    }

    // Inner class for similar books request
    private static class SimilarBooksRequest {

        public Long book_id;
        public Integer top_k;

        public SimilarBooksRequest(Long bookId, Integer topK) {
            this.book_id = bookId;
            this.top_k = topK != null ? topK : 5;
        }
    }
}

File: frontend/src/services/aiService.js
/**
 * AI Service API Client
 * Handles communication with Python AI Service
 */

import axios from 'axios';

// API Base URL - Direct to Python AI service
// Development: localhost, Production: Railway
const API_BASE_URL = process.env.NODE_ENV === 'production' 
  ? 'https://books-store-ai-production.up.railway.app/api'
  : 'http://localhost:8000/api';

/**
 * AI Service API
 */
export const aiService = {
  /**
   * Send chat message to AI
   * @param {string} message - User's message
   * @param {number|null} bookId - Optional book context
   * @param {string|null} category - Optional category filter
   * @param {string|null} sessionId - Optional session ID
   * @returns {Promise} Chat response with answer and sources
   */
  chat: async (message, bookId = null, category = null, sessionId = null) => {
    try {
      const response = await axios.post(`${API_BASE_URL}/chat`, {
        message,
        book_id: bookId,
        category,
        session_id: sessionId
      });
      return response.data;
    } catch (error) {
      console.error('AI Chat error:', error);
      throw error;
    }
  },

  /**
   * Semantic search for books
   * @param {string} query - Search query
   * @param {number} topK - Number of results
   * @param {object} filters - Optional filters (category, minPrice, maxPrice)
   * @returns {Promise} Search results
   */
  search: async (query, topK = 5, filters = {}) => {
    try {
      const response = await axios.post(`${API_BASE_URL}/search`, {
        query,
        topK,
        ...filters
      });
      return response.data;
    } catch (error) {
      console.error('AI Search error:', error);
      throw error;
    }
  },

  /**
   * Get similar books
   * @param {number} bookId - Reference book ID
   * @param {number} topK - Number of similar books
   * @returns {Promise} Similar books
   */
  getSimilarBooks: async (bookId, topK = 5) => {
    try {
      const response = await axios.post(`${API_BASE_URL}/similar`, {
        book_id: bookId,
        top_k: topK
      });
      return response.data;
    } catch (error) {
      console.error('AI Similar books error:', error);
      throw error;
    }
  },

  /**
   * Check AI service health
   * @returns {Promise} Health status
   */
  checkHealth: async () => {
    try {
      const healthUrl = process.env.NODE_ENV === 'production'
        ? 'https://books-store-ai-production.up.railway.app/health'
        : 'http://localhost:8000/health';
      const response = await axios.get(healthUrl);
      return response.data;
    } catch (error) {
      console.error('AI Health check error:', error);
      return { status: 'unavailable' };
    }
  }
};

export default aiService;
